api/handlers.go

```diff
@@ -557,6 +564,122 @@ func (si *ServerImplementation) LookupApplicationByID(ctx echo.Context, applicat
+	if err := si.verifyHandler("LookupApplicationBoxByIDandName", ctx); err != nil {
+		return badRequest(ctx, err.Error())
+	}
+
+	encodedBoxName := params.Name
+	boxNameBytes, err := logic.NewAppCallBytes(encodedBoxName)
+	if err != nil {
+		return badRequest(ctx, err.Error())
+	}
+	boxName, err := boxNameBytes.Raw()
+	if err != nil {
+		return badRequest(ctx, err.Error())
+	}
+
+	q := idb.ApplicationBoxQuery{
+		ApplicationID: applicationID,
+		BoxName:       boxName,
+	}
+	appid, boxes, round, err := si.fetchApplicationBoxes(ctx.Request().Context(), q)
+
+	if err != nil {
+		msg := fmt.Sprintf("%s: round=?=%d, appid=%d, boxName=%s", errFailedLookingUpBoxes, round, applicationID, encodedBoxName)
+		return indexerError(ctx, fmt.Errorf("%s: %w", msg, err))
+	}
+
+	if len(boxes) == 0 {
+		return notFound(ctx, fmt.Sprintf("%s: round=%d, appid=%d, boxName=%+v", errNoApplicationsFound, round, applicationID, encodedBoxName))
+	}
+
+	if appid != generated.ApplicationId(applicationID) {
+		return indexerError(ctx, fmt.Errorf("%s: round=%d, appid=%d, wrong appid=%d, boxName=%s", errWrongAppidFound, round, applicationID, appid, encodedBoxName))
+	}
+
+	if len(boxes) > 1 {
+		return indexerError(ctx, fmt.Errorf("%s: round=%d, appid=%d, boxName=%s", errMultipleBoxes, round, applicationID, encodedBoxName))
+	}
+
+	box := boxes[0]
+	if len(box.Name) == 0 && len(boxName) > 0 {
+		return notFound(ctx, fmt.Sprintf("%s: round=%d, appid=%d, boxName=%+v", errNoBoxesFound, round, applicationID, encodedBoxName))
+	}
+
+	if string(box.Name) != string(boxName) {
+		return indexerError(ctx, fmt.Errorf("%s: round=%d, appid=%d, boxName=%s", errWrongBoxFound, round, applicationID, encodedBoxName))
+	}
+
+	return ctx.JSON(http.StatusOK, generated.BoxResponse(box))
+}
+
+// SearchForApplicationBoxes returns box names for an app
+// (GET /v2/applications/{application-id}/boxes)
+func (si *ServerImplementation) SearchForApplicationBoxes(ctx echo.Context, applicationID uint64, params generated.SearchForApplicationBoxesParams) error {
+	if err := si.verifyHandler("SearchForApplicationBoxes", ctx); err != nil {
+		return badRequest(ctx, err.Error())
+	}
+
+	q := idb.ApplicationBoxQuery{
+		ApplicationID: applicationID,
+		OmitValues:    true,
+	}
+	if params.Limit != nil {
+		q.Limit = *params.Limit
+	}
+	if params.Next != nil {
+		encodedBoxName := *params.Next
+		boxNameBytes, err := logic.NewAppCallBytes(encodedBoxName)
+		if err != nil {
+			return badRequest(ctx, err.Error())
+		}
+		prevBox, err := boxNameBytes.Raw()
+		if err != nil {
+			return badRequest(ctx, err.Error())
+		}
+		q.PrevFinalBox = prevBox
+	}
+
+	appid, boxes, round, err := si.fetchApplicationBoxes(ctx.Request().Context(), q)
+
+	if err != nil {
+		msg := fmt.Sprintf("%s: round=?=%d, appid=%d", errFailedSearchingBoxes, round, applicationID)
+		return indexerError(ctx, fmt.Errorf("%s: %w", msg, err))
+	}
+
```

Comment:
simlarly to the above, query is structured in such a way that:
* no rows back &rarr; no application found
* more than one row back &rarr; the application exists and the box names for the given "page" are provided within the non-nil results

Reply:
I believe the other search endpoints do not return a 404, they would just return an empty array.

Reply:
From a brief investigation, this sounds plausible. For example here [is the app local states handler](https://github.com/algorand/indexer/blob/cf5eb43112f9b0e966c8ea7637a978073ddd7951/api/handlers.go#L275). When no rows are returned, there isn't an error. Even if the lack of rows occurred because a non-existing account id was provided. On the other hand, [when querying for an account id](https://github.com/algorand/indexer/blob/cf5eb43112f9b0e966c8ea7637a978073ddd7951/api/handlers.go#L234) and no rows are returned, there _is a 404 notFound error_. 

I would argue that the more subtle distinction here between non-existing app (404), vs. existing app with non-existing boxes (200) ought to be emulated by **app local states** handler. I.e., just as a 404 is returned when a non-existing account is looked up, a 404 should be returned when local states of a non-existing account is looked up.

So I'm pushing back but am happy to discuss further.

Reply:
Resolving as we orally agreed to continue the distinction that is happening here.